### PR TITLE
Throw runtime exception when a shape or body should have a negative dimension.

### PR DESCRIPTION
**Pull request by peci1**
Meshes are exceptions - you might want to give it a negative scale to mirror it. Although I'm not sure all member functions would tackle negative scale or large negative padding correctly.

### CODE REVIEW

**Comment by peci1** [comment]
I was using random() to generate shapes and negative dimensions slipped my attention. What's interesting is they got through the whole pipeline of shape, body, marker and they even displayed in rviz. But the contains/intersectsRay tests are wrong for such bodies.

File: src/shapes.cpp [hunk 1]
```diff
@@ -285,18 +293,24 @@ void Shape::padd(double padding)
 void Sphere::scaleAndPadd(double scale, double padding)
 {
   radius = radius * scale + padding;
+  if (radius < 0)
```
Comment (rhaschke) [inline src/shapes.cpp]: I think the check for valid dimensions should be made before modifying the object.

**Review by rhaschke — CHANGES_REQUESTED** [review]
I agree that we should check for negative dimensions. In the constructors the checks are fine already.
In the other methods, you should perform the check _before_ modifying the object. Otherwise you leave the object in an undefined state - even if the exception is correctly catched and handled by the caller.

**Comment by peci1** [comment]
I did the requested fixes, please, check if the PR is ready now.

**Review by rhaschke — COMMENTED** [review]
Code looks good. Please rebase onto melodic-devel. We can't accept merge commits in PRs.

**Review by rhaschke — APPROVED** [review]

**Comment by peci1** [comment]
Okay, rebased.

**Comment by v4hn** [comment]
Merging with second approval.
Thanks @peci1 .

**Pull request merged**